cmd/bnsd/x/blueaccount/handler.go

```diff
@@ -0,0 +1,519 @@
+package blueaccount
+
+import (
+	"regexp"
+
+	weave "github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/gconf"
+	"github.com/iov-one/weave/migration"
+	"github.com/iov-one/weave/orm"
+	"github.com/iov-one/weave/x"
+)
+
+func RegisterQuery(qr weave.QueryRouter) {
+	NewDomainBucket().Register("domain", qr)
+	NewAccountBucket().Register("account", qr)
+}
+
+func RegisterRoutes(r weave.Registry, auth x.Authenticator) {
+	r = migration.SchemaMigratingRegistry("blueaccount", r)
+
+	domains := NewDomainBucket()
+	accounts := NewAccountBucket()
+
+	r.Handle(&RegisterDomainMsg{}, &registerDomaiHandler{
+		domains:  domains,
+		accounts: accounts,
+		auth:     auth,
+	})
+	r.Handle(&TransferDomainMsg{}, &transferDomaiHandler{
+		domains: domains,
+		auth:    auth,
+	})
+	r.Handle(&RenewDomainMsg{}, &renewDomaiHandler{
+		domains: domains,
+		auth:    auth,
+	})
+	r.Handle(&DeleteDomainMsg{}, &deleteDomaiHandler{
+		domains: domains,
+		auth:    auth,
+	})
+	r.Handle(&RegisterAccountMsg{}, &registerAccounHandler{
+		auth:     auth,
+		domains:  domains,
+		accounts: accounts,
+	})
+	r.Handle(&TransferAccountMsg{}, &transferAccounHandler{
+		auth:     auth,
+		domains:  domains,
+		accounts: accounts,
+	})
+	r.Handle(&ReplaceAccountTargetsMsg{}, &replaceAccountTargetHandler{
+		auth:     auth,
+		domains:  domains,
+		accounts: accounts,
+	})
+	r.Handle(&DeleteAccountMsg{}, &deleteAccounHandler{
+		auth:     auth,
+		domains:  domains,
+		accounts: accounts,
+	})
+
+	r.Handle(&UpdateConfigurationMsg{}, gconf.NewUpdateConfigurationHandler(
+		"blueaccount", &Configuration{}, auth))
+}
+
+type registerDomaiHandler struct {
+	auth     x.Authenticator
+	domains  orm.ModelBucket
+	accounts orm.ModelBucket
+}
+
+func (h *registerDomaiHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	if _, err := h.validate(ctx, db, tx); err != nil {
+		return nil, err
+	}
+	return &weave.CheckResult{GasAllocated: 0}, nil
+}
+
+func (h *registerDomaiHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	owner := msg.Owner
```

Comment:
```go
owner := msg.Owner
	if len(owner) == 0 {
		owner = x.MainSigner(ctx, h.auth).Address()
	}

	now, err := weave.BlockTime(ctx)
	if err != nil {
		return nil, errors.Wrap(errors.ErrState, "block time not present in context")
	}
	conf, err := loadConf(db)
	if err != nil {
		return nil, errors.Wrap(err, "cannot load configuration")
	}
	domain := Domain{
		Metadata:  &weave.Metadata{},
		Owner:     owner,
		Domain:    msg.Domain,
		ValidTill: weave.AsUnixTime(now.Add(conf.DomainRenew.Duration())),
	}
```

This logic belongs to `validate` method. In this implementation authentication, block time loading config checks are not applied when a tx is received and this could create security vulnurabilty by allowing sending spam tx.
Further read about the topic: 
- https://docs.iov.one/docs/weave/tutorial/handlers#check-vs-deliver
- https://github.com/cosmos/cosmos-sdk/issues/4695

Reply:
I am not sure if this belongs to validation.

`owner` is either `nil` or the main signer. I think this is safe, because there always must be a signer. Otherwise you cannot pay for the transaction.

Block time is always present in the context when running real application, this is assured by Tendermint and user cannot influence this.

Configuration cannot be influenced by the user as well. Invalid configuration is much bigger issue for the application.

Can you explain a bit more what you have in mind?

Reply:
I think we must include @ethanfrey to this conversation, he will enlighten us very well.

Reply:
The use of an optional owner that defaults to MainSigner is a good idiom. You must validate the owner is either a valid address of missing. MainSigner().Address() is guaranteed to be a valid address, as it must have paid a fee.

BlockTime not being in context should never happen unless we change the weave architecture. I actually wanted to make this more explicit: https://github.com/iov-one/weave/issues/822

Reply:
That said, this whole "red pill", "blue pill" issue seems very odd. I will not comment beyond the use of validation logic on a message.

Reply:
By red/blue do you mean the naming?

![redblue](https://user-images.githubusercontent.com/1517/67854955-c0d96800-fb11-11e9-81fd-ffa447c7d1f9.jpg)